pkg/oauthflow/device.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright © 2021 The Sigstore Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oauthflow
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"golang.org/x/oauth2"
+)
+
+const (
+	/* #nosec */
+	SigstoreDeviceURL = "https://oauth2.sigstore.dev/auth/device/code"
+	/* #nosec */
+	SigstoreTokenURL = "https://oauth2.sigstore.dev/auth/device/token"
+)
+
+type deviceResp struct {
+	DeviceCode              string `json:"device_code"`
+	UserCode                string `json:"user_code"`
+	VerificationUri         string `json:"verification_uri"`
+	VerificationUriComplete string `json:"verification_uri_complete"`
+	Interval                int    `json:"interval"`
+}
+
+type tokenResp struct {
+	IDToken string `json:"id_token"`
+	Error   string `json:"error"`
+}
+
+type DeviceFlowTokenGetter struct {
+	MessagePrinter func(string)
+	Sleeper        func(time.Duration)
+	Issuer         string
+	CodeURL        string
+	TokenURL       string
+}
+
+func NewDeviceFlowTokenGetter(issuer, codeURL, tokenURL string) *DeviceFlowTokenGetter {
+	return &DeviceFlowTokenGetter{
+		MessagePrinter: func(s string) { fmt.Println(s) },
+		Sleeper:        time.Sleep,
+		Issuer:         issuer,
+		CodeURL:        codeURL,
+		TokenURL:       tokenURL,
+	}
+}
+
+func (d *DeviceFlowTokenGetter) deviceFlow(clientID string) (string, error) {
+	data := url.Values{
+		"client_id": []string{clientID},
+		"scope":     []string{"openid email"},
+	}
+
+	/* #nosec */
+	resp, err := http.PostForm(d.CodeURL, data)
+	if err != nil {
+		return "", err
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	parsed := deviceResp{}
+	if err := json.Unmarshal(b, &parsed); err != nil {
+		return "", err
+	}
+	d.MessagePrinter(fmt.Sprintf("Open your browser to: %s", parsed.VerificationUriComplete))
```

Comment:
https://tools.ietf.org/html/rfc8628#section-3.2 states that the `verification_uri_complete` field is optional... so might be better to wrap this with a check and substitute `verification_uri` and `user_code` if its not set?

Reply:
Good catch!